### ISSUE TITLE
Contructor injection does noot work with classes

### ISSUE THREAD

**Issue by slim-hmidi** [typestack/typedi#481]
### Description
I have AuthRouter class depends on AuthController which is also depends on AuthService. I inject the dependencies in the constructor of each class but when I call the this.authService inside the AuthController, the injection does not work.

**Minimal code-snippet showcasing the problem**
AuthRouter:
```ts



I create a class AuthRouter:

import { Router } from 'express'
import Container, { Service } from 'typedi'
import AuthenticationController from './index'

@Service()  
class AuthenticationRouter {
  constructor (private readonly authenticationController: AuthenticationController) {}

  getRouter () {
    const router = Router()
    router.get('/auth/url', this.authenticationController.getAuthUrl)

    return router
  }
}

const authRouter = Container.get(AuthenticationRouter)
const routes = authRouter.getRouter()

export default routes

```

AuthController:
```ts
import { Request, Response } from 'express'
import { Service } from 'typedi'
import AuthenticationService from './authentication-service'

@Service()
class AuthenticationController {
  constructor (private readonly authService: AuthenticationService) {}

  async getAuthUrl (req: Request<{}, {}, {}, {redirect: string}>, res: Response) {
    return res.redirect(this.authService.generateAuthenticationUrl(req.query.redirect))
  }

  
}

export default AuthenticationController
```

### Expected behavior

The dependency should be injected correctly by constructor.
### Actual behavior

Dependency does not injected and I got an error:
```

(node:6064) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'authService' of undefined

```

**Comment by desaawa** [comment]
Plus one on that. Just doesn't work as the docs promise.

**Comment by JoaoLSS** [comment]
this has nothing to do with typedi, is about javascript binding, see the error:

`Cannot read property 'authService' of undefined`

this means that `this` is undefined, and that's because when you pass the `getAuthUrl` function to `router.get` it's `this` is not the `AuthenticationController` instance anymore. test this code to see what happens:

```ts
//this class is like the AuthenticationController one
class SomeClass {
    constructor(private readonly someProperty = { someInnerProperty: 'someString' }) {}

    someMethod() {
        console.log(this.someProperty.someInnerProperty)
    }
}

//this class is like the AuthenticationRouter one
class SomeOtherClass {
    constructor(private readonly someClass = new SomeClass()) {}

    someOtherMethod() {
       // im passing a function of the instance "someClass" to setTimeout the same way you're passing a function of the 
       // instance "authenticationController" to router.get
        setTimeout(this.someClass.someMethod, 100)
    }
}

const instance = new SomeOtherClass()
// it will break the exact same way
instance.someOtherMethod()

/***********************************/

//suppose a pseudo implementation of router.get:

router.get(url: string, callback: RequestHandler) {

        // do something

       // here callback doesnt belong to any class, it is just a detached function, and its `this` variable will be the `this` on the
       // scope of router.get  
       callback(req,res,next)
}

//it's like js is doing something like that: 

const callback = this.authenticationController.getAuthUrl
router.get('/auth/url', callback)

// it copies the reference to the function but it doesn't copy the scope where the function is declared
// so to capture the this argument on the function you can do one of two things

//1. use bind() -> this method is available in any function, what it does is that it attaches the value you pass it to the this argument of the function

@Service()
class AuthenticationController {
  constructor (private readonly authService: AuthenticationService) {
     // here im saying that independent of where this function is called the `this` argument should always be this `this`, i.e., the
    // the instance that is been initialized
     this.getAuthUrl = this.getAuthUrl.bind(this) 
  }

  async getAuthUrl (req: Request<{}, {}, {}, {redirect: string}>, res: Response) {
    return res.redirect(this.authService.generateAuthenticationUrl(req.query.redirect))
  }

  
}

//2. use arrow functions -> arrow functions uses closures to keep their scope, so they can mantain the `this` argument independent of where they are called

@Service()
class AuthenticationController {
  constructor (private readonly authService: AuthenticationService) {}

 // here the arrow function is trapping the this of its scope, i.e. the instance of the class where it is defined
 getAuthUrl = async (req: Request<{}, {}, {}, {redirect: string}>, res: Response) => {
    return res.redirect(this.authService.generateAuthenticationUrl(req.query.redirect))
  }

  
}

```

**Comment by slim-hmidi** [comment]
Thanks for your answer and for the details I fixed with another solution:
```
router.get('/google/url', (req: Request<{}, {}, {}, {state: string, code: string}>, res) =>this.authenticationController.getUrl(req, res)
```
 However, I realized that when I added some initialization in the constructor other than the service the attributes are undefined.
E.G:

``` 
//this class is like the AuthenticationController one
class SomeClass {
    pivate attr
    constructor(private readonly someProperty = { someInnerProperty: 'someString' }) {
      this.attr = someValue
    }

    someMethod() {
        console.log(this.someProperty.someInnerProperty)
    }
}

//this class is like the AuthenticationRouter one
class SomeOtherClass {
    constructor(private readonly someClass = new SomeClass()) {}

    someOtherMethod() {
       // im passing a function of the instance "someClass" to setTimeout the same way you're passing a function of the 
       // instance "authenticationController" to router.get
        setTimeout(this.someClass.someMethod, 100)
    }
}
```
`SomeClass`will be instanciated with attr value equals to undefined.

**Comment by JoaoLSS** [comment]
Because on this code the variable `someValue` doesn't exist, so it's value is rightfully `undefined`

**Comment by slim-hmidi** [comment]
I don't understand your answer. Let's be more clear:

```
@Service()
 class TokenService {

  constructor() {
    this.secretKey = "secret"
    }
    
   decode(token){
     jwt.verify(token, this.secretKey)
     }    
}

@Service()
class User {

   constructor(private readonly tokenService: TokenService){}
   
   checkUserToken(data){
      const {token} = data;
      const {userId, tokenId} = tokenService.decode(token);
      // more logic
   }
    
    }
```

When I check the UserToken I got an error mentioned that the secretKey is undefined; so the TokenConstructor does not initialized the secretKey. The only way to make it work is to create a method init which initialize the attributes and call it inside the decode method of the tokenService.

**Comment by JoaoLSS** [comment]
this works just fine, just tested it:

```ts
import 'reflect-metadata';
import { Service, Container } from "typedi";
import jwt from 'jsonwebtoken'

@Service()
 class TokenService {

    private secretKey

  constructor() {
    this.secretKey = "secret"
    }
    
   decode(token: string){
     return jwt.verify(token, this.secretKey)
    }    
}

@Service()
class User {

   constructor(private readonly tokenService: TokenService){}
   
   checkUserToken(data: any){
      const {token} = data;
      const payload = this.tokenService.decode(token);
      console.log({ payload })
   }
    
}

const token = jwt.sign({ userId: 1, tokenId: 2 },"secret")

const instance = Container.get(User)
instance.checkUserToken({ token })
```

**Comment by slim-hmidi** [comment]
My bad, it's due to the env file. I initialized using `process.env.XXX` and when I hardcoded the value it works.
